File: shared/src/main/java/ru/privatenull/pnauth/dialog/DialogType.java
package ru.privatenull.pnauth.dialog;

import java.util.List;

/** Exact type-specific footer/layout variants supported by vanilla Minecraft. */
public sealed interface DialogType permits DialogType.Notice, DialogType.Confirmation,
        DialogType.MultiAction, DialogType.ServerLinks, DialogType.DialogList {
    record Notice(DialogButton action) implements DialogType { }
    record Confirmation(DialogButton yes, DialogButton no) implements DialogType { }
    record MultiAction(List<DialogButton> actions, DialogButton exitAction, int columns) implements DialogType {
        public MultiAction { actions = List.copyOf(actions); }
    }
    record ServerLinks(DialogButton exitAction, int columns, int buttonWidth) implements DialogType { }
    record DialogList(List<PlayerDialog> dialogs, String dialogTag, DialogButton exitAction,
                      int columns, int buttonWidth) implements DialogType {
        public DialogList { dialogs = List.copyOf(dialogs); }
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/dialog/DialogType.kt
package ru.privatenull.pnauth.dialog

/** Exact type-specific footer/layout variants supported by vanilla Minecraft. */
sealed interface DialogType {
    @JvmRecord
    data class Notice(val action: DialogButton) : DialogType

    @JvmRecord
    data class Confirmation(val yes: DialogButton, val no: DialogButton) : DialogType

    @JvmRecord
    data class MultiAction(
        val actions: List<DialogButton>,
        val exitAction: DialogButton?,
        val columns: Int
    ) : DialogType

    @JvmRecord
    data class ServerLinks(val exitAction: DialogButton?, val columns: Int, val buttonWidth: Int) : DialogType

    @JvmRecord
    data class DialogList(
        val dialogs: List<PlayerDialog>,
        val dialogTag: String?,
        val exitAction: DialogButton?,
        val columns: Int,
        val buttonWidth: Int
    ) : DialogType
}
